lib/Dialect/Seq/Transforms/RegOfVecToMem.cpp

```diff
@@ -0,0 +1,245 @@
+  pattern.writeAddr = arrayInject.getIndex();
+  pattern.writeData = arrayInject.getElement();
+  pattern.writeEnable = mux.getCond();
+
+  // Look for read pattern - find array_get users
+  for (auto *user : reg.getResult().getUsers()) {
+    if (auto arrayGet = dyn_cast<hw::ArrayGetOp>(user)) {
+      LLVM_DEBUG(llvm::dbgs() << "  Found array_get: " << arrayGet << "\n");
+      pattern.readAccess = arrayGet;
+      pattern.readAddr = arrayGet.getIndex();
+
+      // Check if read goes through output register
+      for (auto *readUser : arrayGet.getResult().getUsers()) {
+        if (auto outputReg = dyn_cast<FirRegOp>(readUser)) {
+          if (outputReg.getClk() == pattern.clock) {
+            LLVM_DEBUG(llvm::dbgs()
+                       << "  Found output register: " << outputReg << "\n");
+            pattern.outputReg = outputReg;
+            break;
+          }
+        }
+      }
+      break;
+    }
+  }
+
+  bool success = pattern.readAccess != nullptr;
+  LLVM_DEBUG(llvm::dbgs() << "  Pattern analysis "
+                          << (success ? "succeeded" : "failed") << "\n");
+  return success;
+}
+
+bool RegOfVecToMemPass::createFirMemory(MemoryPattern &pattern) {
+  LLVM_DEBUG(llvm::dbgs() << "Creating FirMemory for pattern\n");
+
+  auto dims = getArrayDimensions(pattern.memReg.getType());
+  if (!dims)
+    return false;
+
+  uint64_t depth = dims->first;
+  uint64_t width = dims->second;
+
+  LLVM_DEBUG(llvm::dbgs() << "  Memory dimensions: " << depth << " x " << width
+                          << "\n");
+
+  ImplicitLocOpBuilder builder(pattern.memReg.getLoc(), pattern.memReg);
+
+  // Create FirMem
+  auto memType =
+      FirMemType::get(builder.getContext(), depth, width, /*maskWidth=*/1);
+  auto firMem = builder.create<seq::FirMemOp>(
+      memType, /*readLatency=*/0, /*writeLatency=*/1,
+      /*readUnderWrite=*/seq::RUW::Undefined,
+      /*writeUnderWrite=*/seq::WUW::Undefined,
+      /*name=*/builder.getStringAttr("mem"), /*innerSym=*/hw::InnerSymAttr{},
+      /*init=*/seq::FirMemInitAttr{}, /*prefix=*/StringAttr{},
+      /*outputFile=*/Attribute{});
+
+  // Create read port
+  Value readData = builder.create<FirMemReadOp>(
+      firMem, pattern.readAddr, pattern.clock,
+      /*enable=*/builder.create<hw::ConstantOp>(builder.getI1Type(), 1));
+
+  LLVM_DEBUG(llvm::dbgs() << "  Created read port\n"
+                          << firMem << "\n " << readData);
+
+  Value mask;
+  // Create write port
+  builder.create<FirMemWriteOp>(firMem, pattern.writeAddr, pattern.clock,
+                                pattern.writeEnable, pattern.writeData, mask);
+
+  LLVM_DEBUG(llvm::dbgs() << "  Created write port\n");
+
+  // Replace read access
+  if (pattern.outputReg)
+    // If there's an output register, replace its input
+    pattern.outputReg.getNext().replaceAllUsesWith(readData);
+  else
+    // Replace direct read access
+    pattern.readAccess.getResult().replaceAllUsesWith(readData);
```

Comment:
Does the pass need to distinguish the output register from a regular output read? As in, would it be sufficient to replace all uses of the `array_get` that reads from the memory with the `firmem.read` op?

Reply:
Sounds great :+1: